### ISSUE TITLE
JWT configuration with Firebase (and auth0)?

### ISSUE THREAD

**Issue by sebj54** [mevdschee/php-crud-api#537]
Hi there! 

Thanks for your API helper, it is really useful. I started to use it last week so it's really new to me, I just made it work basically (only CRUD operations for now). I'm actually trying to implement authentification through Firebase with JWT. 

Do you think this will be working? Firebase seems to use JWT (https://firebase.google.com/docs/auth/admin/create-custom-tokens), so I guess it will work but I may be wrong.

First, what is the flow for authentication? I tried to run the auth0 example but did not figured how to make it work…
The token I receive from auth0 does not look to be a valid JWT because it does not have any dot.

I configured my API like this:

```php
$config = new Config([
    'address' => $globalConfig['database']['host'],
    'username' => $globalConfig['database']['username'],
    'password' => $globalConfig['database']['password'],
    'database' => $globalConfig['database']['name'],
    'middlewares' => implode(',', [
        'cors',
        'jwtAuth',
    ]),
    'jwtAuth.secret' => '__MY_SECRET__',
]);
```

I only get errors like this one for each request (which work if I disable jwtAuth middleware):

```json
{
    "code": 1012,
    "message": "Authentication failed for 'JWT'"
}
```
I'd love to provide a working example for JWT with auth0 or Firebase, but I need your help for that 😁 
Thanks in advance!

**Comment by mevdschee** [comment]
Thank you for reporting this issue. Did you see this?

https://github.com/mevdschee/php-crud-api/blob/master/examples/clients/auth0/vanilla.html

Kind regards

**Comment by sebj54** [comment]
Thanks for your quick answer. 

Yes that's exactly what I did to test it out. I copied the `auth0/vanilla.html` file, changed a few variables (`authUrl`, `clientId` & `url`) & ran it in the browser.

The redirection to auth0 works and I can log in with Google. Then I am redirected to my modified `vanilla.html` file. I added a few console logs to show the returned token.

That's this token which seems invalid to me because it has no dots and is quite short (32 characters). When I paste it in the JWT debugger (https://jwt.io/), it is reported as an "invalid signature". 
Any thoughts?

I can provide my auth0 url + client ID if you would like to reproduce.

**Comment by sebj54** [comment]
Just to let you know, I just spent ~ 2 hours on the docs and trying some code, API calls and all and I still can't make it work.

My last try was this sample code (https://auth0.com/docs/quickstart/spa/vanillajs). Even if I get an access token and a JWT token with it, I still have the same error from the API (`1012`).

**Comment by mevdschee** [comment]
I guess this has to do with the auth0 configuration. I'll be happy to help you, but I would need access to your auth0 account. Alternatively you can use this to debug:

https://github.com/mevdschee/php-api-auth

It is a very simple provider that you can run locally.

**Comment by sebj54** [comment]
I made a few tests and I'm close to solve this issue now. 

As I presumed, the error which occurred with the `vanilla.html` file is due to the token format: `access_token` is not a JWT token.
I made a step-by-step debugging to find the cause of the `1012` error code and it is exiting on this line: https://github.com/mevdschee/php-crud-api/blob/master/src/Tqdev/PhpCrudApi/Middleware/JwtAuthMiddleware.php#L26

------

Then, I wanted to call the API with a real JWT token from Auth0, so I used this tutorial: https://auth0.com/docs/quickstart/spa/vanillajs
I made a CodeSandbox for you to test it:

[![Edit Test PHP-CRUD-API JWT](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/static-evmtj?fontsize=14)

Be sure to edit lines 18 & 19 to use your Auth0 account to make the test.

With this test page, the error is the same, but script is now exiting from another line: https://github.com/mevdschee/php-crud-api/blob/master/src/Tqdev/PhpCrudApi/Middleware/JwtAuthMiddleware.php#L60

I also have a warning printed on the page: 

```
Warning:  openssl_verify(): supplied key param cannot be coerced into a public key in [...]\vendor\mevdschee\php-crud-api\src\Tqdev\PhpCrudApi\Middleware\JwtAuthMiddleware.php on line 65
```

It seems that the secret I get from Auth0 is not a real public key. Would a library like this one (https://github.com/firebase/php-jwt) fix the problem?

I don't know if it has to do with the error, but I'm testing on my local Windows machine.

------

Why not giving you access to my account, but my Auth0 account is linked to my Github account and I don't see any way of adding a password to my Auth0 account.

Here is my config anyway:

![image](https://user-images.githubusercontent.com/5880435/58100069-39401280-7bdd-11e9-840d-7f326e46eeae.png)

Let me know if I can help you in any way :)

**Comment by sebj54** [comment]
Ok, after another docs reading session, I found that `audience` parameter is mandatory to get a JWT with the `vanilla.html` file. 
For that, I had to create an API in Auth0 dashboard and add the same URL as the `audience` parameter.

After that, I still had the `openssl_verify` warning. My `secret` was not correct. 
I used the `Client Secret` from the screenshot above, but I should have used the certificate from "Advanced settings" → "Certificates".

So it was totally a PEBKAC :)

Now let's try to make it work with Firebase :sob:
